File: DiscordBotRunner/nodejs-assets/nodejs-project/bots/example-bot.js
'use strict';

// Example Node.js Discord bot.
//
// Contract used by main.js:
//   module.exports.start(ctx) -> optional handle, may be async
//   module.exports.stop(handle) -> optional, may be async
//
// ctx = {token, name, log(line), error(line)}

const {Client, GatewayIntentBits} = require('discord.js');

let client = null;

module.exports = {
  async start(ctx) {
    if (!ctx.token) {
      throw new Error('No bot token supplied. Paste a token in the app.');
    }

    client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
      ],
    });

    client.once('ready', function () {
      ctx.log('Logged in as ' + client.user.tag);
    });

    client.on('messageCreate', async function (message) {
      if (message.author.bot) {
        return;
      }
      if (message.content === '!ping') {
        await message.reply('Pong from a phone!');
      }
    });

    client.on('error', function (err) {
      ctx.error(err.message);
    });

    await client.login(ctx.token);
    return {client: client};
  },

  async stop() {
    if (client) {
      await client.destroy();
      client = null;
    }
  },
};

File: DiscordBotRunner/android/app/src/main/python/bots/example_bot.py
"""Example Python Discord bot (discord.py).

Rename or copy this file to add your own bots. The filename becomes the bot
name shown in the app.
"""

import asyncio

import discord

_client = None
_loop = None


class ExampleClient(discord.Client):
    async def on_ready(self):
        print("Logged in as %s" % self.user)

    async def on_message(self, message):
        if message.author == self.user:
            return
        if message.content == "!ping":
            await message.channel.send("Pong from a phone!")


def start(token, emit):
    global _client, _loop
    if not token:
        raise RuntimeError("No bot token supplied. Paste a token in the app.")
    _loop = asyncio.new_event_loop()
    asyncio.set_event_loop(_loop)
    intents = discord.Intents.default()
    intents.message_content = True
    _client = ExampleClient(intents=intents)
    _loop.run_until_complete(_client.start(token))


def stop():
    global _client, _loop
    if _client is not None and _loop is not None:
        asyncio.run_coroutine_threadsafe(_client.close(), _loop)
